File: Java/src/Intent.java
abstract class Intent {
  private String intentName;
  private String transcript;
  private String[] utterances;

  public Intent(String intentName, String transcript, String[] utterances) {
    this.intentName = intentName;
    this.transcript = transcript;
    this.utterances = utterances;
  }

  public boolean transcriptMatches(String transcript) {
    for (String utterance : this.utterances) {
      if (this.transcript.contains(utterance)) {
        return true;
      }
    }
    return false;
  }

  public abstract String execute();
}

File: JavaScript/Intent.js
class Intent {
  constructor({ intentName, transcript, utterances }) {
    this.intentName = intentName;
    this.transcript = transcript
    this.intentDefinition = {
      utterances
    }
  }

  //parent method that each sub class can safely invoke
  transcriptMatches() {
    for (const utterance of this.intentDefinition.utterances) {
      if (this.transcript.includes(utterance)) {
        return true;
      }
    }
    return false;
  }

  //Method must be overridden in all sub classes
  execute() {
    throw new Error("The execute method has not been defined for this class.");
  }
}

module.exports = { Intent };